docs/lib/routers.md

```diff
@@ -0,0 +1,69 @@
+# Associative Resource Router
+
+django-ansible-base provides an `AssociationResourceRouter` which will auto-construct /associate and /disassociate endpoints for related ManyToMany fields for your models.
+
+To use this router simply do the following:
+```
+from ansible_base.lib.routers import AssociationResourceRouter
+
+router = AssociationResourceRouter()
+
+router.register(
+    r'users',
+    views.UserViewSet,
+    related_views={
+        'teams': (views.TeamViewSet, 'teams'),
+        'organizations': (views.OrganizationViewSet, 'organizations'),
+    },
+)
+
+urlpatterns = [
+    path('api/service/v1/', include(router.urls)),
+]
+```
+
+This would create an endpoint called users with `related` objects teams and organizations. You would also end up with associate/disassociate urls off the related endpoints like:
+  * `/api/service/v1/users/:id/teams/associate`
+  * `/api/service/v1/users/:id/teams/disassociate`
+  * `/api/service/v1/users/:id/organizations/associate`
+  * `/api/service/v1/users/:id/organizations/disassociate`
+
+
+## Breaking down related views
+
+Related views are expressed in the following format:
+
+```
+    '<entry in API related field>': (<ViewSet for relation>, '<related field name>')
+```
+
+NOTE: Often times the `<entry in API related field>` will be the same as `<related field name>` but this is not always the case.
+
+
+If a model does not specify a related field name one will automatically be created by django in the format `<model._meta.model_name>_set`.
+See: https://docs.djangoproject.com/en/dev/topics/db/queries/#backwards-related-objects
+
+For example, in the AuthenticatorMap model we define a foreign key to authenticator like:
```

Comment:
I want to put a pause here. Consider examples up until this point - members of a team, members of an organization. Those are many-to-many. Historically, AWX took relationships like authenticator->authenticator_maps and munged them together with the tooling for many-to-many views. This was a mistake I have tried to correct.

My belief is that a read-only view suits these reverse relationships better. This would require creating a new router factory and having a separate registration mechanism. Whereas m2m is specified as `related_views={`, I would like one-to-many (reverse ForeignKeys) to specify these separately like `reverse_views={`. The reason I want this read-only is because the right way to move an AuthenticatorMap to a new Authenticator is to PATCH to the authenticator_map endpoint `{"authenticator": 2}`, for example. Introducing a new API mechanism introduces new surface area for access control checks, in particular. It has also led to the ability to bypass validation, like moving a host from one inventory to another in AWX (and keeping the same groups, which is deeply wrong).

Reply:
Ok, take a look at the updated code. `related_views` should now happen auto-magically and be read only.

Reply:
Had a look at the code, and it looks great, but wasn't exactly what I expected. I see that `reverse_views` is a separate option that does mostly what I expect. I just don't see any docs or tests around this, and I just want to be sure we're agreed on what the contract looks like.

If you're registering Organization, then you would like "teams" in `reverse_views`. Is that right?

Reply:
Yes, docs added around this. Open to other implementations as well if we want to make this work differently.